File: python/walls_and_gates.py
from collections import deque
from typing import Deque


class Solution:
    """
    @param rooms: m x n 2D grid
    @return: nothing
    """

    def walls_and_gates(self, rooms: list[list[int]]):
        queue: Deque[tuple[int, int]] = deque()
        visited: set[tuple[int, int]] = set()
        for r in range(len(rooms)):
            for c in range(len(rooms[0])):
                if rooms[r][c] == 0:
                    queue.append((r, c))
                    visited.add((r, c))

        distance: int = 0
        while len(queue) > 0:
            for _ in range(len(queue)):
                r, c = queue.popleft()
                rooms[r][c] = distance
                self.walkRoom(r + 1, c, visited, queue, rooms)
                self.walkRoom(r - 1, c, visited, queue, rooms)
                self.walkRoom(r, c + 1, visited, queue, rooms)
                self.walkRoom(r, c - 1, visited, queue, rooms)

            distance += 1

        return

    def walkRoom(
        self,
        row: int,
        col: int,
        visited: set[tuple[int, int]],
        queue: deque[tuple[int, int]],
        rooms: list[list[int]],
    ) -> None:
        if (
            row < 0
            or row >= len(rooms)
            or col < 0
            or col >= len(rooms[0])
            or (row, col) in visited
            or rooms[row][col] == -1
        ):
            return

        visited.add((row, col))
        queue.append((row, col))


def test_calls_and_gates_case_1():
    # arrange
    rooms: list[list[int]] = [
        [2147483647, -1, 0, 2147483647],
        [2147483647, 2147483647, 2147483647, -1],
        [2147483647, -1, 2147483647, -1],
        [0, -1, 2147483647, 2147483647],
    ]

    expected: list[list[int]] = [
        [3, -1, 0, 1],
        [2, 2, 1, -1],
        [1, -1, 2, -1],
        [0, -1, 3, 4],
    ]

    # act
    solution = Solution()
    solution.walls_and_gates(rooms)

    # assert
    assert expected == rooms


def test_calls_and_gates_case_2():
    # arrange
    rooms: list[list[int]] = [[0, -1], [2147483647, 2147483647]]

    expected: list[list[int]] = [[0, -1], [1, 2]]

    # act
    solution = Solution()
    solution.walls_and_gates(rooms)

    # assert
    assert expected == rooms

File: rust/src/code/walls_and_gates.rs
use std::collections::{HashSet, VecDeque};

struct Solution {}

type ROW = usize;
type COL = usize;

// NOTE: Time complexity: O(n * m)
impl Solution {
    pub fn walls_and_gates(mut rooms: &mut Vec<Vec<i32>>) {
        let mut queue: VecDeque<(ROW, COL)> = VecDeque::new();
        let mut visited: HashSet<(ROW, COL)> = HashSet::new();
        for row in 0..rooms.len() {
            for col in 0..rooms[0].len() {
                if rooms[row][col] == 0 {
                    queue.push_back((row, col));
                    visited.insert((row, col));
                }
            }
        }

        let mut distance: i32 = 0;
        while !queue.is_empty() {
            for _ in 0..queue.len() {
                if let Some((r, c)) = queue.pop_front() {
                    rooms[r][c] = distance;
                    Self::walk_room(r + 1, c, &mut queue, &mut rooms, &mut visited);
                    if r.checked_sub(1).is_some() {
                        Self::walk_room(r - 1, c, &mut queue, &mut rooms, &mut visited);
                    }

                    Self::walk_room(r, c + 1, &mut queue, &mut rooms, &mut visited);
                    if c.checked_sub(1).is_some() {
                        Self::walk_room(r, c - 1, &mut queue, &mut rooms, &mut visited);
                    }
                }
            }
            distance += 1;
        }
    }

    fn walk_room(
        row: usize,
        col: usize,
        queue: &mut VecDeque<(ROW, COL)>,
        rooms: &mut Vec<Vec<i32>>,
        visited: &mut HashSet<(ROW, COL)>,
    ) {
        if row >= rooms.len()
            || col >= rooms[0].len()
            || visited.contains(&(row, col))
            || rooms[row][col] == -1
        {
            return;
        }
        visited.insert((row, col));
        queue.push_back((row, col));
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn walls_and_gates_case_1() {
        // arrange
        let mut rooms: Vec<Vec<i32>> = vec![
            vec![2147483647, -1, 0, 2147483647],
            vec![2147483647, 2147483647, 2147483647, -1],
            vec![2147483647, -1, 2147483647, -1],
            vec![0, -1, 2147483647, 2147483647],
        ];

        let expected = vec![
            vec![3, -1, 0, 1],
            vec![2, 2, 1, -1],
            vec![1, -1, 2, -1],
            vec![0, -1, 3, 4],
        ];

        // act
        Solution::walls_and_gates(&mut rooms);

        // assert
        assert_eq!(expected, rooms);
    }

    #[test]
    fn walls_and_gates_case_2() {
        // arrange
        let mut rooms: Vec<Vec<i32>> = vec![vec![0, -1], vec![2147483647, 2147483647]];
        let expected = vec![vec![0, -1], vec![1, 2]];

        // act
        Solution::walls_and_gates(&mut rooms);

        // assert
        assert_eq!(expected, rooms);
    }
}
